compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala

```diff
@@ -0,0 +1,425 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import Decorators._, Flags._, Types._, Contexts._, Symbols._, Constants._
+import ast.Trees._
+import ast.{TreeTypeMap, untpd}
+import util.Positions._
+import tasty.TreePickler.Hole
+import SymUtils._
+import NameKinds._
+import dotty.tools.dotc.ast.tpd
+import typer.Implicits.SearchFailureType
+
+import scala.collection.mutable
+import dotty.tools.dotc.core.StdNames._
+import dotty.tools.dotc.core.quoted._
+import dotty.tools.dotc.typer.Inliner
+import dotty.tools.dotc.util.SourcePosition
+
+
+/** Translates quoted terms and types to `unpickle` method calls.
+ *
+ *  Transforms top level quote
+ *   ```
+ *   '{ ...
+ *      val x1 = ???
+ *      val x2 = ???
+ *      ...
+ *      ~{ ... '{ ... x1 ... x2 ...} ... }
+ *      ...
+ *    }
+ *    ```
+ *  to
+ *    ```
+ *     unpickle(
+ *       [[ // PICKLED TASTY
+ *         ...
+ *         val x1 = ???
+ *         val x2 = ???
+ *         ...
+ *         Hole(0 | x1, x2)
+ *         ...
+ *       ]],
+ *       List(
+ *         (args: Seq[Any]) => {
+ *           val x1$1 = args(0).asInstanceOf[Expr[T]]
+ *           val x2$1 = args(1).asInstanceOf[Expr[T]] // can be asInstanceOf[Type[T]]
+ *           ...
+ *           { ... '{ ... x1$1.unary_~ ... x2$1.unary_~ ...} ... }
+ *         }
+ *       )
+ *     )
+ *    ```
+ *  and then performs the same transformation on `'{ ... x1$1.unary_~ ... x2$1.unary_~ ...}`.
+ *
+ */
+class ReifyQuotes extends MacroTransformWithImplicits {
+  import tpd._
+  import ReifyQuotes._
+
+  override def phaseName: String = ReifyQuotes.name
+
+  override def checkPostCondition(tree: Tree)(implicit ctx: Context): Unit = {
+    tree match {
+      case tree: RefTree if !ctx.inInlineMethod =>
+        assert(!tree.symbol.isQuote)
+        // assert(!tree.symbol.isSplice) // TODO widen ~ type references at stage 0?
+        assert(tree.symbol != defn.QuotedExpr_~)
+      case tree: Select if tree.symbol == defn.QuotedExpr_~ =>
+        assert(Splicer.canBeSpliced(tree.qualifier))
+      case _ =>
+    }
+  }
+
+  override def run(implicit ctx: Context): Unit =
```

Comment:
I think you should override `isRunnable` and not `run`:
```scala
override def isRunnable(implicit ctx: Context): Boolean =
  super.isRunnable && ctx.compilationUnit.needsStaging
```

Reply:
Actually I am wrong. `isRunnable` is for the whole phase and not a single compilation unit